CCZ hosting revamp

mkangia opened the pull request:
[Spec](https://docs.google.com/document/d/1pIST_6VfEPflG09x8R8Btt0a8-FjRgWVW8MHlMR_Gkk/edit#heading=h.rv7nxn3p5lc2)

[Slightly more specific Spec](https://docs.google.com/document/d/1pIST_6VfEPflG09x8R8Btt0a8-FjRgWVW8MHlMR_Gkk/edit#heading=h.ekamn78h2ds)

Currently the file hosted via nginx roughly looks like
https://github.com/dimagi/icds-files/blob/master/hindi_files/index.html

This PR adds ability to configure CCZs, supporting files, page title and CCZ file names which should make it possible to achieve similar structure as the page earlier.
My effort is to code it in a way that it can be easily extracted out of ICDS in case product decides to make it a part of the product. That did drive some decisions.

Please reach out to me if you need to do a walk through/demo.
I am trying to push this to staging so it can be viewed directly there as well.
Here is a slightly long [demo video](http://recordit.co/53VYQM8YK8).
The CCZ creation is handled by celery, the task was kept running in the background during the demo.

P.S: a link to visit the hosting page was added to the CCZ hosting listing after the demo in [commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/91cfe286c22b36f26917eac5a3ff274a68cc7f45)

Might be easier to review all at once but the commits should give a nice breakdown as well

HTML will be updated to 2 spaces before merge, if it has not already.

[Working example](https://staging.commcarehq.org/a/myfirsttestproject/ccz/hostings/manage)
password is 123456

mkangia committed:
improvement: avoid iteration

mkangia committed:
rename

mkangia committed:
add models

mkangia committed:
add model validation

mkangia committed:
add CCZHostingLinks interface

mkangia committed:
fix error due to inherited view

mkangia committed:
raise more specific exception

mkangia committed:
update verion search to set initial value and only show released versions

mkangia committed:
add CCZHosting mgmt interface

mkangia committed:
add bare ccz hosting view with authentication behaviour

mkangia committed:
add link under application tab, CCZHostingtab and urls

mkangia committed:
fix property call

mkangia committed:
lint

mkangia committed:
merge master

mkangia committed:
update according to js

mkangia committed:
extract create_ccz_files and break out other methods

mkangia committed:
add missing imports

mkangia committed:
allow adding in a name when storing new file

mkangia committed:
add task to set up ccz for hosting

mkangia committed:
list out all app versions on hosting page

mkangia committed:
delete file from blobdb on delete

mkangia committed:
lint

mkangia committed:
fix name

mkangia committed:
delete ccz on delete/update if not needed anymore

mkangia commented on custom/icds/tasks.py:
```diff
@@ -9,6 +9,8 @@
 from corehq.sql_db.util import get_db_aliases_for_partitioned_query
 from corehq.util.datadog.gauges import datadog_counter
+# import tasks to get discovered by autodiscover_tasks
+from custom.icds.tasks import *
```
this is needed.

esoergel replied:
@mkangia what's the deal with this?  I've never seen a file import from itself - what does it accomplish?

esoergel replied:
Oh, I see - you've made a directory called `tasks/`, but not removed `tasks.py`.  Can you reconcile that?  I'm surprised that that works, but it seems ambiguous.  Looks like there's only one thing in there, which you could move into the `__init__.py` file

mkangia replied:
Ya i wanted to set things up this way. so a directory for tasks and then each task should have its own file.
I was not sure if i should just remove the tasks.py.
you are saying i should move that to tasks/__init__ instead? that sounds good to me.
[commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/869cca27227949f92011d2ac704949cd9fee979f)

mkangia commented on custom/icds/tasks/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from __future__ import absolute_import
+
+from custom.icds.tasks.ccz_hosting import *
```
this is needed

mkangia replied:
and then add each task here when we add a new one?

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/24129/commits/29ca21f9a3a72f24dd08be6ba561b9feb84a1f81

mkangia committed:
allow toggle for user as well

mkangia committed:
populate build profiles

mkangia committed:
add profile_id to model

mkangia committed:
replace update_version with simple save

- make each entry unique person link, app, version, profile_id
- remove update_version to avoid complexity managing with profile_id and make things more flexible to let user configure multiple version per profile or multiple profile per version for an app

mkangia committed:
list profile name in list

mkangia committed:
use profile id when creating ccz

mkangia committed:
fix download link

mkangia committed:
better alignment

mkangia committed:
parse version to int

mkangia committed:
migrate last few changes

mkangia committed:
add ability to set ccz file name

mkangia committed:
some improvements

mkangia committed:
add footer for ICDS envs

mkangia committed:
add page title

mkangia committed:
set profile id in link when searching

mkangia committed:
set build profile initial value after search/on page load

mkangia committed:
spacing

mkangia committed:
stickler

mkangia committed:
stickler

mkangia committed:
add ccz_file_meta to IcdsFile instead

mkangia committed:
update bootstrap

mkangia committed:
add supporting files

mkangia committed:
better page alignment

mkangia committed:
lint

mkangia committed:
validate that build is present

mkangia committed:
adapt delete ccz with profile_id

mkangia committed:
remove reduntant domain context

mkangia committed:
fetch build only when necessary

mkangia committed:
add tests

mkangia committed:
lint

mkangia committed:
setup supporting docs just above the footer

mkangia committed:
Merge branch 'master' into mk/ccz-hosting-revamp

mkangia committed:
fix search

mkangia committed:
render options in case of search

mkangia committed:
add link

mkangia committed:
better check

mkangia committed:
add ability to delete a link

mkangia committed:
call delete on associated model coz cascade does not

mkangia committed:
open in new tab

snopoke commented on settings.py:
```diff
@@ -895,6 +895,12 @@
 
 ES_SETTINGS = None
 
+CCZ_FILE_HOSTING_SUPPORTING_FILES = {
```
not sure how I feel about having this in settings. Is there not a way to get them out the blob db directly without having to list them here? If you set the blob parent to the `CCZHostingLink` then it should be easy enough to list them all right?

mkangia replied:
these files are shared by all links in a domain.
The process as of now is,
setup a file in blob db from the backend at the server and then make it available for HQ to show for all links in the domain. I thought id was the safest way to map it correctly and since this does not need flexibility i.e they are not going to be changed frequently (as of day they would change just when LTS updates) putting this in settings seemed the right choice without adding another model/complexity.

In a follow up some time down the line, we can consider making this configurable from the HQ interface itself where one can upload a file, give it a name, mark it as a supporting file/document and thats it. but as of now i thought it was better to just keep it simple.

mkangia replied:
These are now removed and moved to a model

snopoke reviewed (commented):
Generally OK with this but a few notable comments:
* password should not be stored in the DB but rather the hashed password
* Use of `IcdsFile` should be re-considered
* Storing DB IDs in settings should be re-considered

mkangia committed:
check for build profile only when needed

mkangia committed:
hack for null value in js for blank option

esoergel commented on custom/icds/tests/models/test_ccz_hosting.py:
```diff
@@ -0,0 +1,81 @@
+)
+
+
+@mock.patch('custom.icds.tasks.setup_ccz_file_for_hosting.delay')
```
FWIW you can pass `MagicMock()` here as an arg so it doesn't pass the default mock object to your methods

esoergel replied:
Oh, I see you did that for some tests later

mkangia replied:
i think i simply ignored the mocks i didn't care for by `*_`. An object would be created anyway right?

esoergel replied:
Yeah, it's the same except for whether the mocks get passed to the function.

esoergel commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -68,3 +77,67 @@ def get_context_data(self, **kwargs):
+        if self.form.is_valid():
+            success, error_message = self.form.save()
+            if success:
+                return redirect(self.urlname, domain=self.domain)
```
That's pretty typical - it's usually done so reloading the page doesn't submit the form again

esoergel commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -0,0 +1,267 @@
+
+
+@location_safe
+def download_ccz(request, domain, hosting_id, blob_id):
```
I assume this view and the following one intentionally don't require auth, is that correct?

mkangia replied:
ya those have been intentionally skipped. 
Though i should try adding the same authentication as the parent view.

esoergel commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -0,0 +1,267 @@
+)
+from corehq.apps.hqwebapp.decorators import use_select2_v4
+from corehq.apps.locations.permissions import location_safe
+from corehq.apps.domain.decorators import login_and_domain_required
```
`login_and_domain_required` seems insufficient for these views - shouldn't it check that the user is a domain admin, say?

mkangia replied:
i would leave that to GS feedback. there was a concern raised when the FF was just at domain level, to which i changed it to allow at user level as well.
Domain admin is a stronger check but from past experience the team has asked for special permission to just limited users configured instead of giving it to all admins.
So giving user based access on FF looked like the quick fix without introducing more complexity.

esoergel replied:
Does that mean dropping the domain level flag too?  With the domain flag turned on, any user on the domain will be able to access these views.  I could also see this switching to "either domain admin or feature flag", but that feels more complicated to implement unless necessary.

esoergel reviewed (commented):
I still need to review the JS and HTML, but it looks good overall.  The demo is very helpful, thanks for doing that again.

mkangia commented on custom/icds/models.py:
```diff
@@ -0,0 +1,102 @@
+                                  validators=[LowercaseAlphanumbericValidator])
+    username = models.CharField(null=False, max_length=255)
+    # b64_aes_encrypt'ed raw password is stored in DB
+    password = models.CharField(null=False, max_length=255)
```
While developing/testing i found it useful to see what the password is to a link. If we keep it one way we would need to keep storing passwords in keepasses again whereas this makes it possible to have things setup just on HQ. I am not sure if we consider this bad security because these are not user accounts so at some level this should be okay to be visible on HQ.
We have used the same implementation at more places in HQ so i just built the same.
Open to reconsideration if this is still concerning

mkangia replied:
Updated to store passwords as salted hashed password.

mkangia commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -471,11 +474,10 @@ def get_available_versions_for_app(domain, app_id):
-    raise Exception("Build for version requested not found")
+    build = get_build_doc_by_version(domain, app_id, version, return_doc=False)
+    if not build:
+        raise Exception(_("Build for version requested not found"))
```
this was updated.

mkangia commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -108,16 +109,18 @@ def get_latest_build_version(domain, app_id):
 
 
-def get_build_doc_by_version(domain, app_id, version):
+def get_build_doc_by_version(domain, app_id, version, return_doc=True):
```
in `get_version_build_id`. The method added in this PR that does not need the doc

mkangia commented on custom/icds/templates/icds/manage_ccz_hosting_links.html:
```diff
@@ -0,0 +1,44 @@
+            <ul data-bind="foreach: links">
+                <li>
+                    <a data-bind="attr: { href: editUrl}, text: identifier"></a>
+                    <a data-bind="attr: { href: pageUrl}">
```
Definitely. Updating.

mkangia replied:
put this on the wrong link, this is later updated on the correct link pageUrl

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/24129/commits/652eebe8e99b0555e11c51f6a5d08ca300ae752a

mkangia commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -0,0 +1,70 @@
+
+
+@location_safe
+@method_decorator([toggles.APP_TRANSLATIONS_WITH_TRANSIFEX.required_decorator()], name='dispatch')
```
This was corrected later. I remember i left a comment on the commit about this correction so that the reviewer does get notified too. Not sure if that was not shown.

mkangia commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -0,0 +1,70 @@
+            'domain': self.domain,
+        }
+
+    @method_decorator(login_and_domain_required)
```
None. Updating. Thanks.

mkangia commented on custom/icds/forms.py:
```diff
@@ -34,3 +45,60 @@ def __init__(self, *args, **kwargs):
+    def clean(self):
+        if self.cleaned_data.get('app_id') and self.cleaned_data.get('version'):
+            try:
+                self.version_build_id
```
i think i updated that at some point.. no wonder pycharm was highlighting it. Thanks.

mkangia commented on custom/icds/models.py:
```diff
@@ -38,9 +38,24 @@ class CCZHosting(models.Model):
         return get_build_by_version(self.link.domain, self.app_id, self.version)
 
+    @classmethod
+    def update_version(cls, link_id, app_id, version):
```
this method was later removed and i used simple model save.

But to answer your question, if i remember it correctly it was to call `full_clean` because its not called otherwise i.e related to validations added on the model that need to be checked.

mkangia commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -68,3 +77,67 @@ def get_context_data(self, **kwargs):
+    template_name = 'icds/manage_ccz_hosting.html'
+    section_name = ugettext_noop('CCZ Hostings')
+
+    @use_select2_v4
```
i will try that.. i think that fails, i have tried that earlier but didn't dig into the reason for it.

mkangia replied:
had similar issues with this. Won't be getting into debugging that, so keeping it like this.

mkangia commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -68,3 +77,67 @@ def get_context_data(self, **kwargs):
+    def get_context_data(self, **kwargs):
+        app_names = {app.id: app.name for app in get_brief_apps_in_domain(self.domain, include_remote=True)}
+        if self.request.GET.get('link_id'):
+            ccz_hostings = CCZHosting.objects.filter(link_id=self.request.GET.get('link_id'))
```
it already filters by link id which should be unique to a domain?

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/24129/commits/a86738b2b221458154e0481d7b2981337907b995

mkangia commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -68,3 +77,67 @@ def get_context_data(self, **kwargs):
+@login_and_domain_required
+def remove_ccz_hosting(request, domain, link_id):
+    try:
+        ccz_hosting_link = CCZHosting.objects.get(pk=link_id)
```
a better check, cool.

mkangia commented on custom/icds/tasks/ccz_hosting.py:
```diff
@@ -0,0 +1,35 @@
+    ccz_blob_id = ccz_hosting.blob_id
+    build = wrap_app(get_build_by_version(ccz_hosting.link.domain, ccz_hosting.app_id, version))
+    # set up the file if not already present
+    if not IcdsFile.objects.filter(blob_id=ccz_blob_id, data_type="ccz").exists():
```
`IcdsFile` was mentioned on the spec doc when i shared that earlier for tech review. But it definitely makes sense to add a new utility instead specially about the expiry, we need these files to persist. thanks.

mkangia replied:
IcdsFile use removed. Updated CCZHostingUtility to handle file management

mkangia commented on custom/icds/templates/icds/ccz_hosting.html:
```diff
@@ -34,6 +34,11 @@ <h3>
         <div class="row">
             <div class="col-sm-1"></div>
             <div class="col-sm-9">
+                {% if icds_env %}
```
thats a fair point. Just added this in case this is used by any other project/domain/environment like staging, though i could not get this to work on staging because icds-ucr is not configured so the testing might happen on india.

mkangia commented on settings.py:
```diff
@@ -895,6 +895,12 @@
+    # domain: { 'file name in listing': 'blob id' }
+    # 'mkangia-domain': {'Terms of Service / Licensing': '01c3f9ced57f477ea1c826e48bca7eed'}
+}
+CCZ_FILE_HOSTING_SUPPORTING_DOCS = {}
```
There are two sets of supporting things
1. files which are the apks and are listed just like the cczs
2. the second set is supporting documents like license, how to install etc which are listed in a different way just above the footer.
So this is the latter.

mkangia replied:
these are now removed.

mkangia commented on corehq/apps/hqmedia/tasks.py:
```diff
@@ -149,81 +137,126 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
+    fpath = create_ccz_files(app, build_profile_id, include_multimedia_files, include_index_files,
+                             download_id, compress_zip, filename, download_targeted_version,
+                             track_progress=True, expose_link=True)
+    _expose_download_link(fpath, filename, compress_zip, download_id)
```
most of it would need to adapt to changes made by Jenny in another PR. 
Tried to make the splits as logical as possible.

orangejenny replied:
Agreed, this is a good split. And we are likely going to have merge conflicts, yes, but I don't think they'll be major.

mkangia replied:
@orangejenny i will try to merge before you :)

orangejenny replied:
🏎 🏎

mkangia commented on corehq/apps/hqmedia/tasks.py:
```diff
@@ -149,81 +137,126 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
+                     download_id=None, compress_zip=False, filename="commcare.zip",
+                     download_targeted_version=False, track_progress=False, expose_link=False):
+    """
+    :param track_progress: track progress of the files when being run asynchronously by celery
```
I am scared about making changes to this code. I just kept it in the same order and specifics as it was earlier.

esoergel replied:
splitting it up into multiple functions changes the equation a bit though - previously, only code inside `build_application_zip` referenced `build_application_zip`.  Now `create_ccz_files` and `zip_ccs_files` both reference that function, meaning that they're coupled to it. 

`build_application_zip` is being passed to `set_progress` as the way to communicate to celery what task is being processed.  For instance, how does `create_ccz_files` know that it's being called from `build_application_zip`?

mkangia replied:
@esoergel 
i am not fully sure how the tracking of the task would change with these changes. I just tested it once and saw the progress bar updated but i can't say it worked exactly the way its suppose to
Are you saying the tracking would not work correctly now?

snopoke replied:
Agree with @esoergel that instead of passing a boolean you should just pass the task function itself in which case you just change:
```diff
- if track_progress:
-     DownloadBase.set_progress(build_application_zip, current_progress, 100)
DownloadBase.set_progress(celery_task, current_progress, 100)
```

You don't even have to check if it's `None` since that's handled in `set_progress`

mkangia commented on custom/icds/models.py:
```diff
@@ -0,0 +1,107 @@
+    page_title = models.CharField(blank=True, max_length=255)
+
+    def __str__(self):
+        return self.identifier
```
yes, thanks for the correction.
Thats what i intended but did it incorrectly.

mkangia replied:
i take that back. i needed __str__ only so that in the views/html it's rendered like that. __repr__ didn't do that for me.

esoergel replied:
Yeah, `__repr__` is just for poking around in the command line and similar.  You might also consider using an explicit `render` method or something if this is intended to be shown to the user, otherwise it can be hard to trace down where this is used if you wanted to change it later.

esoergel replied:
Oh, or pass `obj.identifier` to the template when it's going to be rendered rather than `obj`

mkangia replied:
okay. removed it and used identifier.
I was using it to populate choices for link on `CCZHostingForm`

mkangia commented on custom/icds/models.py:
```diff
@@ -0,0 +1,107 @@
+        super(CCZHostingLink, self).delete(*args, **kwargs)
+
+
+class CCZHosting(models.Model):
```
i wanted it to be different from any vocabulary used already to avoid confusion so just `CCZ` would beat that.
`UploadedCCZ` does seem right because the CCZ is not being uploaded (anymore :), thats the aim of this PR ).
CCZs are being hosted for use, hence CCZHosting.
Makes sense?
(tbh, i personally don't love that name but could not come up with anything better)

esoergel replied:
Maybe `HostedCCZ`?  I try to think of model names as a simple descriptor of one row in the database - as in "this user created a new X".  I guess `CCZHosting` works in that case too.  Whatever you prefer.

orangejenny replied:
Agree `HostedCCZ` would be better, if only because it's more obviously a noun.

mkangia replied:
renamed this particular model [here](https://github.com/dimagi/commcare-hq/pull/24129/commits/6782754a1139e3e832a2d12043cbf44f1bf3acac)

Left other models/views as-is. Hopefully this wont be confusing now since model is HostedCCZ and other models/views are like `CCZHostingLink`, `CCZHostingSupportingFile`, `ManageCCZHosting` or probably i just got accustomed to everything starting with `CCZHosting`

millerdev replied:
Would be ideal to change all instances of `CCZHosting` to `HostedCCZ` for consistency.

- `CCZHostingSupportingFile` -> `HostedCCZSupportingFile`
- `ManageCCZHosting` -> `ManageHostedCCZ`

mkangia replied:
okay. Updated in [commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/23ce553c113c4aa36e5e239cf11283ff519f937f). Only the toggle says CCZ Hosting which seemed fine.

mkangia commented on corehq/toggles.py:
```diff
@@ -1758,3 +1758,11 @@ def _commtrackify(domain_name, toggle_is_enabled):
+    'Allow project to configure ccz hosting',
+    TAG_CUSTOM,
+    [NAMESPACE_USER, NAMESPACE_DOMAIN]
+)
```
Since this has `NAMESPACE_USER` always_enabled would not be correct right?

In terms of environment i think it would be needed on india and icds and may be staging. I think i am fine to keep it open to other environments.

esoergel replied:
The "enabled" checks look for a user OR domain match.  If it's going to be enabled for a domain, then `always_enabled` should work fine for that, and wouldn't preclude users from using it on other domains.

orangejenny replied:
Does this need to be both user and domain? I thought we were moving away from user-oriented flags?

mkangia replied:
@esoergel 
> If it's going to be enabled for a domain

Ya thats why i am more keen on keeping it per user because if it's set for a domain, all users would start seeing the option to manage it and we wouldn't want that.

mkangia replied:
@orangejenny 
> I thought we were moving away from user-oriented flags

Is it? User specific permissions should be handled via role/permissions?

orangejenny replied:
Domain-oriented flags are generally better because they tell us more about where a feature is in use. But if this needs to be a user-specific feature, feature flag is fine, I don't think you need to muck around in user permissions.

esoergel replied:
Definitely agree that this shouldn't touch user permissions.  Can you drop the `NAMESPACE_DOMAIN`, then?  It sounds like turning this on for a whole domain would be a bad idea, especially since the views are open to any user on the domain as discussed [here](https://github.com/dimagi/commcare-hq/pull/24129/files#r281403565)

mkangia replied:
@esoergel updated to keep just as user
https://github.com/dimagi/commcare-hq/pull/24129/commits/f7a63c0219825eab0c5c4e9f2cd2aa7b51ba5269

mkangia commented on custom/icds/forms.py:
```diff
@@ -0,0 +1,140 @@
+    def clean_link_id(self):
+        link_id = self.cleaned_data.get('link_id')
+        if not link_id:
+            self.add_error('link_id', _("Please select link"))
```
Its because the way form is structured. The same form handles the search which does not needed these fields to be required.

esoergel replied:
I don't follow - wouldn't that usage of the form also hit the same `clean` methods?

mkangia replied:
The form is used for search as well in which case these fields are not required. If these are marked as required, either the form would not submit or the form would show errors on search which we dont want. So the validation are added separately when we manually check for form being valid.

esoergel replied:
Do you mean on the client side?  What I'm saying is that it looks like these methods make these fields required, even for search - wouldn't there need to be something like `if is_search: return link_id` to bypass this validation?

mkangia replied:
not sure what changed but now things work well without them too.
So removing these three methods till i can justify having them.
[commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/edf5a035295031864d930dfc117fc3fe736639a3)

mkangia committed:
decorators on class, domain checks, fix method call

mkangia commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -472,11 +477,10 @@ def get_available_versions_for_app(domain, app_id):
 
 
 def get_version_build_id(domain, app_id, version):
```
If you the final implementation of `get_version_build_id`, it passes return_doc=False to `get_version_build_id` so it should not get the doc but just the result doc(or whatever the correct term for the mapped minimal result is for couch views)
Does that sound right to you?

esoergel replied:
Ah, you're right - I misread that.  I didn't see that `include_docs` isn't provided if `not return_doc`

mkangia committed:
better logic + name for checking if version exists

mkangia commented on custom/icds/migrations/0005_cczhostinglink_page_title.py:
```diff
@@ -0,0 +1,21 @@
+            name='page_title',
+            field=models.CharField(blank=True, max_length=255),
+        ),
+    ]
```
Is it better to just delete the migrations and create a new one right before merging?
i can just rollback on local environment and can clean up staging

snopoke replied:
yes, that's what I mean

mkangia committed:
only import the specific task

mkangia commented:
@snopoke 
Thanks for the feedback
Addressed review in https://github.com/dimagi/commcare-hq/pull/24129/commits/d076888e668eae61ea495240cacfc17094ab7ab8 and https://github.com/dimagi/commcare-hq/pull/24129/commits/29ca21f9a3a72f24dd08be6ba561b9feb84a1f81

Will look into making a custom blob file handler and removing use of IcdsFile.
And 
https://github.com/dimagi/commcare-hq/pull/24129#discussion_r281139270
`I think domain should be populated with the current domain and not specified on the UI`

rest is addressed.

mkangia commented:
@esoergel 
thanks for the feedback. Have addressed review in https://github.com/dimagi/commcare-hq/pull/24129/commits/1da909d8e1d3b7e991378c399ba4c4b4a97fd255 and left comments/replies to rest.

mkangia committed:
assign domain from the view itself

mkangia commented:
@snopoke @esoergel 
I should have mentioned this before,
my effort was to code this in a way that it can be easily extracted out of ICDS in case product decides to make it a part of the product. That did drive some decisions.

mkangia committed:
merge icds_file_obj and ccz_file_blob

mkangia committed:
put target on correct link

mkangia committed:
pass required argument to form

mkangia committed:
update CCZHostingUtility and use that instead of IcdsFile

mkangia commented:
@snopoke 
#24129 (comment)
> I think domain should be populated with the current domain and not specified on the UI

https://github.com/dimagi/commcare-hq/pull/24129/commits/d9368fdabc66dc96c6d433f34dad75690b470724 and followup https://github.com/dimagi/commcare-hq/pull/24129/commits/6206256c36c28a01c590a6e6d10d55d7ce29283e

Removed use of IcdsFile https://github.com/dimagi/commcare-hq/pull/24129/commits/4b5eecd4e9068f09dd20503c40643fce52dc2f31

All review points were addressed.

mkangia committed:
lint

mkangia committed:
fix tests

orangejenny commented on corehq/apps/app_manager/dbaccessors.py:
```diff
@@ -109,16 +110,20 @@ def get_latest_build_version(domain, app_id):
     return res['value']['version'] if res else None
 
 
-def get_build_doc_by_version(domain, app_id, version):
```
Another nitpick: I'd split this into two functions, one that calls the other, rather than have a parameter that changes a function's return value.

mkangia replied:
@orangejenny @esoergel 
my intention was to rewrite and rename as to say
get build doc and return the result or return the doc if this argument is true. Hence the rename to skip doc and the argument `return_doc`, can also update to `return_as_doc` if that makes more sense.

> one that calls the other
I would not be able to control the kwargs in that case. the overwrite lets me skip getting the doc at all if i don't need it.

snopoke replied:
+1 for @orangejenny's suggestion to have 2 functions. They can both call a 3rd private function that does the DB query

mkangia replied:
I did the [revert of the commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/a7310be0c3de204eff4b72138a631b8542f434eb) and then added the private method [here](https://github.com/dimagi/commcare-hq/pull/24129/commits/2fb510e1879de97b667170254a1018bef3e72bde)

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/widgets_v4.js:
```diff
@@ -53,12 +54,33 @@ hqDefine("app_manager/js/widgets_v4", [
+                var buildProfiles = _.map(e.params.data.buildProfiles, function (details, profileId) {
+                    return { id: profileId, text: details.name };
+                });
+                buildProfiles = Array.prototype.concat({id: 'null', text: gettext("Select profile")},
```
Use a placeholder to do this: https://select2.org/placeholders

orangejenny commented on corehq/apps/app_manager/static/app_manager/js/widgets_v4.js:
```diff
@@ -53,12 +54,33 @@ hqDefine("app_manager/js/widgets_v4", [
         }
+        var appProfileInitialValue = initialPageData.get('appProfileInitialValue');
+        if (!_.isEmpty(appProfileInitialValue)) {
+            $('#app-profile-id-input').select2({data: appProfileInitialValue});
```
Doesn't the input need to be transformed to a select2 regardless of whether or not there's data?

mkangia replied:
If there is no initial value, there is no need to set anything. The select options will populate when the user selects the version. Hence this was skipped.

orangejenny commented on corehq/apps/hqmedia/tasks.py:
```diff
@@ -149,81 +137,126 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
+    return files, errors, file_count
+
+
+def _zip_ccs_files(fpath, files, current_progress, file_progress, file_count, compression, track_progress):
```
CCS files? Should this be CCZ?

mkangia replied:
same as https://github.com/dimagi/commcare-hq/pull/24129#discussion_r281763571

orangejenny replied:
Not following - like "zip cc's files"? Thinking about the other comment, how about `_zip_files_for_ccz`?

snopoke replied:
+1 for `_zip_files_for_ccz`

orangejenny commented on corehq/apps/hqmedia/tasks.py:
```diff
@@ -149,81 +137,126 @@ def build_application_zip(include_multimedia_files, include_index_files, app,
+                    DownloadBase.set_progress(build_application_zip, current_progress, 100)
+
+
+def create_ccz_files(build, build_profile_id, include_multimedia_files=True, include_index_files=True,
```
Why is this `files`, plural? Looks like it still just creates one.

mkangia replied:
i saw this as create files for the ccz and those are multiple.
do you suggest just `create_ccz` ?

orangejenny replied:
How about `create_files_for_ccz`?

orangejenny commented on custom/icds/forms.py:
```diff
@@ -0,0 +1,141 @@
+
+    def clean_profile_id(self):
+        profile_id = self.cleaned_data.get('profile_id')
+        # ugly hack for the null added in the javascript
```
Is this because of the null that could be a placeholder, or the null added in the "clear all options manually" section (or both)?

mkangia replied:
this would be because of the placeholder that i had to put. I see you have suggested a way to add that without setting the null value. So if i get that working, this would not be needed anymore.

orangejenny commented on custom/icds/models.py:
```diff
@@ -0,0 +1,107 @@
+    file_name = models.CharField(max_length=255, blank=True)
+
+    class Meta:
+        unique_together = ('link', 'app_id', 'version', 'profile_id')
```
Why would you have multiple links for the same app + version + build profile combination?

orangejenny replied:
To phrase another way, why does `link` need to be in this list, shouldn't the other three alone be unique?

mkangia replied:
@orangejenny 
We can have the same version and profile on an app on two links like two states using the same language/profile and on the same version of the app.

orangejenny commented on custom/icds/models.py:
```diff
@@ -0,0 +1,107 @@
+    def blob_id(self):
+        assert self.app_id
+        assert self.version
+        return "%s%s%s" % (self.app_id, self.version, self.profile_id)
```
I'd have expected some kind of delimiter, like `"%s-%s-%s"`, both for readability while debugging and for the very unlikely case that app_id+version+profile_id. Or is this style typical of blob ids?

mkangia replied:
thats not a typical style of blob ids. This is generated specifically for this only.
I dont think one should ever need to use this to deduce the app_id, version and profile_id since that can be seen on the object itself. I am fine with keeping this as a string of characters that look like an id.

orangejenny replied:
okay, that sounds fine

millerdev replied:
Not a big deal, but `"%s-%s-%s"` would be a fine format for such a long and complex multipart blob id.

mkangia commented on custom/icds/urls.py:
```diff
@@ -0,0 +1,29 @@
+        name=EditCCZHostingLink.urlname),
+    url(r'^ccz/hostings/link/(?P<hosting_id>[\d-]+)/delete/', remove_ccz_hosting,
+        name="remove_ccz_hosting"),
+    url(r'^ccz/hostings/(?P<hosting_id>[\w-]+)/download/(?P<blob_id>[\w-]+)/', download_ccz,
```
Sounds good. i see that blob id is not needed even now since we have the hosting_id. Though it's needed for the supporting docs/files.
(yet to look at the other comment)

mkangia replied:
The blob id was removed from the url and the view. Just hosting id was sufficient

mkangia commented on custom/icds/utils/ccz_hosting.py:
```diff
@@ -0,0 +1,62 @@
+    def ccz_details(self):
+        file_name = self.ccz_hosting.file_name
+        if not file_name and self.get_file_meta:
+            file_name = self.get_file_meta.name
```
makes sense, thanks.

orangejenny commented on custom/icds/templates/icds/manage_ccz_hosting_links.html:
```diff
@@ -0,0 +1,44 @@
+                <li>
+                    <a data-bind="attr: { href: editUrl}, text: identifier"></a>
+                    <a data-bind="attr: { href: pageUrl}" target="_blank">
+                        <span class="fa fa-arrow-right"></span>
```
I don't think it's clear that clicking the name means edit and then the icon is a separate link that goes to the page. I'd do the names in plain text and then display two buttons or links afterwards, one for "edit" and one for "view"/"open". The scheduled reports UI does something like this.

For icons, there's a `fa-external-link` that more commonly means "this will open in a new tab" if you don't like the eye icon that's used for the view button in scheduled reports.

mkangia replied:
added `fa-external-link` but could not add button for edit in a way that makes it look nicer.

orangejenny replied:
Okay. Given how few users this page will have, I'm fine with calling that out of scope and revisiting if people find it hard to use.

orangejenny commented on custom/icds/validators.py:
```diff
@@ -0,0 +1,12 @@
+from django.utils.translation import ugettext_lazy as _
+
+
+LowercaseAlphanumbericValidator = RegexValidator(
```
nitpick: "alphanumeric" not "alphanumberic"

orangejenny commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -0,0 +1,265 @@
+            except CCZHostingLink.DoesNotExist:
+                pass
+            else:
+                ccz_hosting_link.delete()
```
Shouldn't this line go in the `try`?

mkangia replied:
You mean a new try block?
fwiw, i dont expect `delete` to fail.

orangejenny replied:
Ignore this, I was confused about exactly what `else` does. Your code is good.

orangejenny commented on custom/icds/views/ccz_hosting.py:
```diff
@@ -0,0 +1,265 @@
+        except CCZHostingLink.DoesNotExist:
+            return HttpResponse(status=404)
+
+        uname, passwd = get_username_and_password_from_request(request)
```
Nitpick: `username` and `password` are much more standard names.

orangejenny commented:
Reviewed and left a few minor comments, nothing major. Not marking approved because there's more substantial feedback from other people that looks like it needs to be addressed.

mkangia committed:
better naming [skip-ci]

mkangia committed:
use placeholder to avoid using "null" [skip-ci]

mkangia committed:
better listing of links

mkangia committed:
store name always on ccz_hosting, fallback on meta only for supporting docs/files that don't have ccz hosting

mkangia committed:
fix tests

mkangia commented on custom/icds/utils/ccz_hosting.py:
```diff
@@ -0,0 +1,62 @@
+        return get_blob_db().size(key=self.blob_id)
+
+    @cached_property
+    def get_file_meta(self):
```
the reason i keep `get_` in the name because the object is getting it or obtaining it from another place and is not set on the object itself and hence there is some db call or call on another object is going to happen. Just keeping it a little explicit.
i think setting it cached_property then looks incorrect and should be just memoized.

millerdev replied:
I agree, it would be better to use `memoized` decorator rather than `cached_property`.

mkangia committed:
use memoized instead to respect naming

millerdev commented on custom/icds/tasks/ccz_hosting.py:
```diff
@@ -0,0 +1,31 @@
+        except Exception as e:
+            # delete the file from blob db if it was added but later failed
+            ccz_utility.remove_file_from_blobdb()
+            raise e
```
This will not preserve the stack trace. Suggestion: use [`six.reraise`](https://pythonhosted.org/six/#six.reraise)
```py
except Exception:
    exc = sys.exc_info()
    ccz_utility.remove_file_from_blobdb()
    six.reraise(*exc)

mkangia replied:
updated.

millerdev commented on custom/icds/utils/ccz_hosting.py:
```diff
@@ -0,0 +1,62 @@
+    def store_file_in_blobdb(self, ccz_file, name=None):
+        db = get_blob_db()
+        try:
+            kw = {"meta": db.metadb.get(parent_id='CCZHosting', key=self.blob_id)}
```
Maybe I suggested too many improvements in one comment. However, I think elimination of `settings.CCZ_FILE_HOSTING_SUPPORTING_FILES` should not be dismissed so lightly:

    # Storing the blob_id and domain in CCZHosting eliminates the need for
    # settings.CCZ_FILE_HOSTING_SUPPORTING_FILES. All of the information
    # that was stored in settings is now available in the CCZHosting table.

I still think it would be good to add a `domain` field to `CCZHosting` so that those `settings` values can be eliminated (@snopoke also [requested something similar](https://github.com/dimagi/commcare-hq/pull/24129/files#r281156258)).

---

You've pointed out in an offline chat that the `blob_id` is a derived field:

https://github.com/dimagi/commcare-hq/blob/5f113267e72f0bf735bf626d3fc68a607e838b78/custom/icds/models.py#L65-L69

and that key uniquely identifies a given blob, and the content of that blob will never change (blob ids are not reused). If all that is true, then it seems reasonable to use a composite/calculated `blob_id` as defined. Additionally, if all that is true then the code of the method on which we are commenting is more complex than necessary:

```py
db.metadb.get(parent_id='CCZHosting', key=self.blob_id)
```

will always raise `BlobMeta.DoesNotExist` (blob ids are not reused), and therefore it would be better to remove the `try:` block and change this method to do:

```py
db.put(
    ccz_file,
    domain=self.ccz_hosting.link.domain,
    # varying parent_id means the metadata records will be distributed
    # across all shards of the metadata database.
    parent_id="CCZHosting-%s" % self.ccz_hosting.id,
    type_code=CODES.tempfile,
    key=self.blob_id,
    name=name,
)
```

If an error is raised due to an existing blob that would be overwritten then there is an error in the code or in your assumptions about the uniqueness of `blob_id`, and we want to know about that.

mkangia replied:
@millerdev 
If i have the same version of "app_id + version + profile" for two links, this would fail when i would try to save blobmeta.
Even if i was to get pass that, in task `setup_ccz_file_for_hosting` that does the setup, i would always need to generate the ccz to pass to this method.

Working on removing things from settings as CCZHosting objects.
One thing i would need to do is to either
1. remove the validations on model for app_id and version since it would need to be blank for the supporting apps and docs.
2. or set up value like "dummy" for app_id and 0 for version for these docs. [I would prefer this one because i would like to keep the validations on the model]

mkangia replied:
Update:
I added a new model instead for supporting files instead of making this complex.

millerdev replied:
I'm not sure how to convince you anymore. If line 48 does not raise `BlobMeta.DoesNotExist`

https://github.com/dimagi/commcare-hq/blob/e2c9f5ad2b0f4e1260ea5e4f5389fb79b55aa19d/custom/icds/utils/ccz_hosting.py#L47-L49

 then this code is reusing the blob id (that is, writing new content to an existing blob key), which you previously said will never happen.

mkangia replied:
I hear you.
>However, I think elimination of settings.CCZ_FILE_HOSTING_SUPPORTING_FILES should not be dismissed so lightly

i made this change.

> will always raise BlobMeta.DoesNotExist (blob ids are not reused)

I see that check is confusing, removed it [here](https://github.com/dimagi/commcare-hq/pull/24129/commits/7155265a6b6f46fc09657052dc4552fb37dc7861) since we would not restore an already present blob

Reiterating on the why i cant use "CCZHosting-%s" % self.ccz_hosting.id" with the same blob id since that is not allowed due to unique constraint on blob id(key) for blob meta

mkangia committed:
use reraise to preserver stacktrace

mkangia committed:
fix method call

mkangia committed:
only delete if not needed, hence delete via ccz_hosting

mkangia committed:
minor improvements

mkangia committed:
add CCZHositingSupportingFile

mkangia committed:
use property itself

mkangia committed:
adapt utility for CCZHosting/CCZHostingSupoprtingFiles

mkangia committed:
remove blob id and add fallback for file name

mkangia committed:
add an early check

mkangia committed:
add create for easier setup

mkangia committed:
store one way salted hashed password

mkangia committed:
remove unused imports

mkangia committed:
adapt and fix test with recent changes

mkangia committed:
remove __str__ overrite

mkangia committed:
lint and settings removed

mkangia committed:
filter by domain too

mkangia committed:
remove blob id from url, its not needed in view anymore

mkangia committed:
revert changes made to IcdsFile

mkangia commented:
Hey
@esoergel @snopoke @orangejenny @millerdev 

Thank you for keeping up with this.
I believe i have addressed all blockers.

This PR is getting pretty huge so i would like to get this out if there are no blockers and take care of any improvements in follow up PRs for easier review.
The feedback that i have addressed, i have marked those comments as resolved for compact view and keeping pending comments still revealed.

Majorly
1. updated passwords to be salted hash. Currently they can't be updated from the interface, just added on create.
2. Moved supporting files/docs from settings to a separate model instead. This would make it easier in case we need to add a HQ interface depending on the usage of the feature.

Everything else post your last review should be improvement or other review feedback.

mkangia committed:
fix url

mkangia commented:
[working example](https://staging.commcarehq.org/a/myfirsttestproject/ccz/hostings/manage)

mkangia committed:
remove redundant test since we removed __str__ overrite

mkangia committed:
fix test with url changes

mkangia committed:
ensure profile id None if not set

mkangia committed:
remove unused import

snopoke commented on custom/icds/models.py:
```diff
@@ -0,0 +1,160 @@
+
+    def delete(self, *args, **kwargs):
+        for ccz_hosting in CCZHosting.objects.filter(link=self):
+            ccz_hosting.delete()
```
is this not handled by `on_delete=CASCADE` on the `CCZHosting` foreign key?

mkangia replied:
that would not call the delete method on CCZHosting but just delete the object.
I updated to use signal instead of that its not confusing
https://github.com/dimagi/commcare-hq/pull/24129/commits/2b905b38c2199971ad033d106199cbbd4088327e

millerdev commented on custom/icds/tasks/ccz_hosting.py:
```diff
@@ -28,4 +30,4 @@ def setup_ccz_file_for_hosting(ccz_hosting_id):
             # delete the file from blob db if it was added but later failed
             ccz_utility.remove_file_from_blobdb()
-            raise e
+            six.reraise(*sys.exc_info())
```
It is important to capture `sys.exc_info()` immediately inside the `except` block since `ccz_hosting.delete_ccz()` might change the exception context. Example:
```py
>>> import sys
>>> import six

>>> # bad re-raise exception
... try:
...     raise Exception("should re-raise")
... except:
...     try:
...         raise Exception("should NOT re-raise")
...     except:
...         pass
...     six.reraise(*sys.exc_info())
...
Traceback (most recent call last):
  File "<stdin>", line 9, in <module>
  File "<stdin>", line 6, in <module>
Exception: should NOT re-raise

>>> # good re-raise exception
... try:
...     raise Exception("should re-raise")
... except:
...     exc = sys.exc_info()
...     try:
...         raise Exception("should NOT re-raise")
...     except:
...         pass
...     six.reraise(*exc)
...
Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "<stdin>", line 3, in <module>
Exception: should re-raise
```

mkangia replied:
https://github.com/dimagi/commcare-hq/pull/24129/commits/f80a2641320dab474b2e70cc1c59a30d571c3ef6

mkangia committed:
reserve stacktrace

mkangia committed:
use signal instead of overriting delete to remove cczs

mkangia committed:
let the parent method set task kwarg and use that instead

mkangia committed:
rename to _zip_files_for_ccz and create_files_for_ccz

mkangia committed:
Revert "rename"

This reverts commit dba263c38957d1a5e6837480e19a07c7ba451054.

mkangia committed:
remove unused import

mkangia committed:
use a private method instead of overriding get_buld_doc_by_version

mkangia committed:
rename other instances of get_build_by_version added after the reverted commit

mkangia committed:
fix test

mkangia committed:
Merge branch 'master' into mk/ccz-hosting-revamp

mkangia committed:
blob is not reused, so the check is irrelevant and should never be needed

mkangia committed:
squash migrations

mkangia commented:
@orangejenny @esoergel @snopoke @millerdev 

responded/made changes for the recent comments.

summary
- renaming of methods, create_ccz_files and _zip_ccz_files
- not passing global `build_application_zip` when tracking progress
- squash migrations
- use signal instead of overriting delete method on CCZHostinglink
- better changes for get_build_doc_by_version improvement
- since blob id is not reused, don't check for blobmeta presence when storing ccz file

mkangia committed:
move tasks.py into tasks/__init__

mkangia committed:
indent 2 spaces, align closing tags to opening —skip-ci--

mkangia commented on corehq/apps/hqmedia/tasks.py:
```diff
@@ -190,24 +190,24 @@ def _zip_ccs_files(fpath, files, current_progress, file_progress, file_count, co
                 current_progress += file_progress / file_count
-                if track_progress:
-                    DownloadBase.set_progress(build_application_zip, current_progress, 100)
+                if task:
```
@esoergel yes simon pointed that earlier.
should not we avoid few method calls if we can? So i already know that if task is not present, why set progress?

esoergel replied:
Method calls have so little overhead in Python that it's not worth factoring in as a performance consideration.  I'd make the decision entirely based on what makes the cleanest code.  I think it's cleaner to avoid checking whether `task` exists every time it's referenced, considering that `task=None` is an accepted part of the API.  You're certainly free to disagree with that opinion, however.

mkangia replied:
okay, updated.

mkangia committed:
remove toggle at domain level —skip-ci—

mkangia committed:
remove clean_* methods till they can be justified

mkangia committed:
add import —skip-ci—

mkangia committed:
lint —skip-ci—

mkangia committed:
Merge branch 'master' into mk/ccz-hosting-revamp

mkangia committed:
let set_progress handle None task

millerdev commented on custom/icds/models.py:
```diff
@@ -0,0 +1,164 @@
+    identifier = models.CharField(null=False, unique=True, max_length=255, db_index=True,
+                                  validators=[LowercaseAlphanumericValidator])
+    username = models.CharField(null=False, max_length=255)
+    # b64_aes_encrypt'ed raw password is stored in DB
```
Is this comment still relevant? According to https://github.com/dimagi/commcare-hq/pull/24129/files#r283154584 it is now salted and hashed.

mkangia replied:
It's not. removed.
Thanks

millerdev commented on custom/icds/utils/ccz_hosting.py:
```diff
@@ -0,0 +1,55 @@
+    def store_file_in_blobdb(self, file_obj, name):
+        kw = {
+            "domain": self.file_hosting.domain,
+            "parent_id": 'CCZHosting',
```
From a previous thread, which I can no longer reply to
> Reiterating on the why i cant use "CCZHosting-%s" % self.ccz_hosting.id" with the same blob id since that is not allowed due to unique constraint on blob id(key) for blob meta

I'm not following your reasoning for why `"CCZHosting-%s" % self.ccz_hosting.id"` will not work; there is no unique constraint on `parent_id`. Although I do see `self.file_hosting` is not always a `CCZHosting` object anymore.

The issue is that using a constant `parent_id` will cause metadata for all these blobs to be saved in the same database shard. That might not be a problem if there are only a few of them. But if it is expected that there will be many then it would be better to use a `parent_id` value that will cause the metadata to be distributed across shards. I do not know the best decision here since I don't know the expected count of these blobs.

Maybe `"%s-%s" % (type(self.file_hosting).__name__, self.file_hosting.id)` would work? `parent_id=self.blob_id` would certainly work, although that would make it harder to find the parent object given only the blob metadata.

mkangia replied:
> there is no unique constraint on `parent_id`
@millerdev there is a unique constraint on `key` on BlobMeta and so i can't add blob meta again for a new ccz hosting in case the ccz for the same app + version + build profile is already present.

mkangia replied:
> That might not be a problem if there are only a few of them

i am guessing if we consider only ICDS, and they launch in all 29 states with each link having the 3 applications or even if they follow some other convention i believe this would be < 100.

millerdev replied:
> @millerdev there is a unique constraint on key on BlobMeta and so i can't add blob meta again for a new ccz hosting in case the ccz for the same app + version + build profile is already present.

There must be excactly one `BlobMeta` record per blob, no more and no less—the design of the blob db requires it. The issue you raise here should not be a concern since you are not reusing blob keys and it's not possible to have two blobs with the same key.

millerdev replied:
> ... i believe this would be < 100

Yes, that's a very small number. Distribution among shards is not a concern if that is the max. We would need to revisit this decision if that ever changes.

proteusvacuum committed:
Merge branch 'master' into mk/ccz-hosting-revamp

proteusvacuum commented:
@mkangia I merged master so I could deploy staging, but not totally sure if merge was the right thing. Could you check to make sure?

mkangia commented:
hey @proteusvacuum 
that does not look right, the code you added should go in `_build_ccz_files`, some part of it is already there.

mkangia committed:
fix merge

- adapt to https://github.com/dimagi/commcare-hq/pull/24303/files
- add file_cache when zipping files for ccz
- remove _ensure_all_media_files_in_ccz as it was replaced by check_ccz_multimedia_integrity
- remove code repeateing for building ccz files

mkangia commented:
i fixed the merge
@orangejenny can you help confirm it, the conflict was with your code changes in https://github.com/dimagi/commcare-hq/pull/24303/files

mkangia committed:
remove missed if task check

Commit:
Merge branch 'master' into mk/ccz-hosting-revamp

orangejenny commented:
@mkangia I never quite know how to verify that a merge conflict was handled correctly, but I looked over the `hqmedia.tasks` changes and they look fine. The tests added in https://github.com/dimagi/commcare-hq/pull/24303/ still pass, so that gives me confidence.

mkangia commented:
okay, thanks @orangejenny

mkangia committed:
remove redundant comment [skip ci]

mkangia committed:
CCZHosting -> HostedCCZ, model and serializer

mkangia commented:
@orangejenny @esoergel @snopoke @millerdev 
Can i get a "approved" on this if there is no blocking comment?

mkangia committed:
rename all ccz_hosting to hosted_ccz

mkangia committed:
lint [skip ci]

esoergel committed:
select2 v4 is now the default

https://github.com/dimagi/commcare-hq/pull/24365

millerdev reviewed (commented):
👍 on ✅ for the blob db-related code. Will let others sign off on the other parts.

esoergel reviewed (approved):
I think all the feedback has been addressed.  Thanks for being diligent in following up, @mkangia

mkangia commented:
Thank you @millerdev  and @esoergel 
I don't think the build failure is related so restarted that job

mkangia committed:
merge master

mkangia committed:
Remove unused import --skip ci--

snopoke commented on custom/icds/models.py:
```diff
@@ -0,0 +1,163 @@
+        hosted_ccz.delete_ccz()
+
+
+pre_delete.connect(delete_ccz_for_link, sender=HostedCCZLink)
```
I commented on this before but can't find the comment now. Is this not handled by `on_delete=CASCADE` of the foreign key?

mkangia replied:
@snopoke 
on_delete=CASCADE does not call the delete method on the related object but just deteles it directly from DB. We need some clean up to be done when the related object is deleted.

Here is the [commit](https://github.com/dimagi/commcare-hq/pull/24129/commits/2b905b38c2199971ad033d106199cbbd4088327e) where i switched to use signal instead of overriting the delete method on CCZHostingLink after your earlier comment

mkangia replied:
here is your comment from before https://github.com/dimagi/commcare-hq/pull/24129#discussion_r283323558

snopoke reviewed (approved):
Migrations still need to be merged into one.

mkangia commented:
Offlined with @snopoke about the migrations and keeping it as-is since this has already been run on staging.

mkangia commented:
fyi @orangejenny 
did one final conflict resolution [here](https://github.com/dimagi/commcare-hq/pull/24129/commits/f6a214b29862507004764458033a2fa5adac1d01)

mkangia commented:
Here it goes!
